Separate prototyping and production concerns in any-kubernetes-cluster.md (#3363)
knative/docs: docs/install/any-kubernetes-cluster.md

You can install Knative by applying YAML files using the `kubectl` CLI.
You can install the Serving component, Eventing component, or both on your cluster.

## System Minimum Requirements

> For prototyping purposes, Knative can work on local deployments of Kubernetes (minikube, kind, etc.).

Reviewer: ```suggestion
For prototyping purposes, Knative can work on local deployments of Kubernetes.
```
Wouldn't mention these by name because people then start asking for specific docs for them and we don't do per-platform docs.

Author: I think it's important to mention them by name... when people ask we will say "no"

Reviewer: Also, this is a bit confusing now I think because we don't give min requirements for the local deployments, just the production deployments? Should we even be saying production purposes before we're calling everything v 1.0? cc @n3wscott @evankanderson

Author: I don't mind removing these bits altogether, my core concern is not having system requirements that are beyond people's local machines at the top of the Install doc (I don't want to scare people)

Reviewer: What's the point not telling them the truth though? If the requirements are beyond the scope of what they have access to we can't really do anything about that, why would we want someone to start using it anyway if their machine can't handle it and eventually it just won't work and will have wasted their time? These need to be verified properly anyway but that to me would be better, i.e, make them accurate instead of just removing them.

Also, I disagree about it being important to mention minikube and kind, it's just going to annoy people if you tell them they can use those but that we're not going to produce docs for it, and it mixes the message we were trying to send by removing platform specific docs in the first place IMO.

Author: My only concern is making sure people know that a prototype of Knative will work on their local machine... 

>If the requirements are beyond the scope of what they have access to we can't really do anything about that, why would we want someone to start using it anyway if their machine can't handle it and eventually it just won't work and will have wasted their time?

We aren't saying "you can run a production application with the resources of your laptop" but we do want to say "hey you can try this out with basically any K8s deployment." 

>Also, I disagree about it being important to mention minikube and kind, it's just going to annoy people if you tell them they can use those but that we're not going to produce docs for it, and it mixes the message we were trying to send by removing platform specific docs in the first place IMO.

I yield on this point 👍🏼

For production purposes, we recommend:
- If you have only one node in your cluster, you will need at least 6 CPUs, 6 GB of memory, and 30 GB of disk storage.

> - If you have multiple nodes in your cluster, for each node you will need at least 2 CPUs, 4 GB of memory, and 20 GB of disk storage.

Reviewer 2: So , I just set up a `kind` cluster with a single node with 2 CPU and 4GB RAM using Kourier (cribbed from https://github.com/csantanapr/knative-kind):

```shell
$ kubectl describe nodes
Name:               kind-control-plane
Roles:              master
...
Capacity:
  cpu:                2
  memory:             4042468Ki
  pods:               110
Allocatable:
  cpu:                2
  ephemeral-storage:  103211296Ki
  memory:             4042468Ki
  pods:               110
Non-terminated Pods:          (23 in total)
  Namespace                   Name                                          CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
  ---------                   ----                                          ------------  ----------  ---------------  -------------  ---
  default                     hello-00001-deployment-75c8ff6d45-gs64n       25m (1%)      0 (0%)      0 (0%)           0 (0%)         14s
  knative-eventing            eventing-controller-d666b4657-7jdw6           100m (5%)     0 (0%)      100Mi (2%)       0 (0%)         2m21s
  knative-eventing            eventing-webhook-778b6b8cf4-mg8vn             20m (1%)      200m (10%)  20Mi (0%)        200Mi (5%)     2m21s
  knative-eventing            imc-controller-5f4bdf86cf-gt97m               0 (0%)        0 (0%)      0 (0%)           0 (0%)         111s
  knative-eventing            imc-dispatcher-54bfc97957-rvt2c               0 (0%)        0 (0%)      0 (0%)           0 (0%)         111s
  knative-eventing            mt-broker-controller-ff696f56c-k8zfg          100m (5%)     0 (0%)      100Mi (2%)       0 (0%)         103s
  knative-eventing            mt-broker-filter-d44b776d8-4sq75              100m (5%)     0 (0%)      100Mi (2%)       0 (0%)         103s
  knative-eventing            mt-broker-ingress-6c8487b74c-gh2fm            100m (5%)     0 (0%)      100Mi (2%)       0 (0%)         103s
  knative-serving             3scale-kourier-control-547fc6c565-4rspk       0 (0%)        0 (0%)      0 (0%)           0 (0%)         5m12s
  knative-serving             activator-86956bbd6f-qvrpn                    300m (15%)    1 (50%)     60Mi (1%)        600Mi (15%)    5m23s
  knative-serving             autoscaler-54cbd576f6-vj28h                   30m (1%)      300m (15%)  40Mi (1%)        400Mi (10%)    5m23s
  knative-serving             controller-79c9cccd6f-2rsxp                   100m (5%)     1 (50%)     100Mi (2%)       1000Mi (25%)   5m23s
  knative-serving             webhook-5fd484cf4-xl5c5                       100m (5%)     500m (25%)  100Mi (2%)       500Mi (12%)    5m23s
  kourier-system              3scale-kourier-gateway-5665d74bd4-zvxxb       0 (0%)        0 (0%)      0 (0%)           0 (0%)         5m12s
  kube-system                 coredns-f9fd979d6-g96x6                       100m (5%)     0 (0%)      70Mi (1%)        170Mi (4%)     5m52s
  kube-system                 coredns-f9fd979d6-pfvk7                       100m (5%)     0 (0%)      70Mi (1%)        170Mi (4%)     5m52s
  kube-system                 etcd-kind-control-plane                       0 (0%)        0 (0%)      0 (0%)           0 (0%)         5m51s
  kube-system                 kindnet-6z76p                                 100m (5%)     100m (5%)   50Mi (1%)        50Mi (1%)      5m52s
  kube-system                 kube-apiserver-kind-control-plane             250m (12%)    0 (0%)      0 (0%)           0 (0%)         5m51s
  kube-system                 kube-controller-manager-kind-control-plane    200m (10%)    0 (0%)      0 (0%)           0 (0%)         5m51s
  kube-system                 kube-proxy-d4l7l                              0 (0%)        0 (0%)      0 (0%)           0 (0%)         5m52s
  kube-system                 kube-scheduler-kind-control-plane             100m (5%)     0 (0%)      0 (0%)           0 (0%)         5m51s
  local-path-storage          local-path-provisioner-78776bfc44-r247s       0 (0%)        0 (0%)      0 (0%)           0 (0%)         5m51s
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests     Limits
  --------           --------     ------
  cpu                1825m (91%)  3100m (155%)
  memory             910Mi (23%)  3090Mi (78%)
  ephemeral-storage  0 (0%)       0 (0%)
```

CPU gets close at 2 CPUs with one user-container running (requests is near capacity, limits is over), but I think we could say that 3 CPUs and 5GB RAM is a lower bound for reasonable experimentation.

I'll also point out this is with `kind` requesting almost 1 CPU (2*0.1+0.1+0.25+0.2+0.1 = 0.85)

You can install Knative by applying YAML files using the `kubectl` CLI.
You can install the Serving component, Eventing component, or both on your cluster.

> ## Recommended System Requirement

Reviewer: ```suggestion
## Recommended system requirements
```

Reviewer: I don't think we need to say recommended? Not sure if this was part of the discussion I missed though

> For prototyping purposes, Knative will work on most local deployments of Kubernetes. For example, you shouldn't have any issues spinning up a local one node clister with 2 CPU and 4GB of memory.

Reviewer: ```suggestion
For prototyping purposes, Knative will work on most local deployments of Kubernetes. For example, you can use a local, one-node cluster that has 2 CPU and 4GB of memory.
```

> For production purposes, we recommend:
> - If you have only one node in your cluster, you will need at least 6 CPUs, 6 GB of memory, and 30 GB of disk storage.
> - If you have multiple nodes in your cluster, for each node you will need at least 2 CPUs, 4 GB of memory, and 20 GB of disk storage.

Reviewer: ```suggestion
For production purposes, it is recommended that:
- If you have only one node in your cluster, you will need at least 6 CPUs, 6 GB of memory, and 30 GB of disk storage.
- If you have multiple nodes in your cluster, for each node you will need at least 2 CPUs, 4 GB of memory, and 20 GB of disk storage.
```

> <!--TODO: Verify these requirements-->

Reviewer: Are these verified now?

> These "Recommended System Requirements" are just that, recommendations. Your mileage may vary depending on your choice of Networking layer, whether or not you use a service mesh, etc.

Reviewer: ```suggestion
**NOTE:** The system requirements provided are recommendations only. The requirements for your installation may vary, depending on whether you use optional components, such as a networking layer.
```
